### DISCUSSION TITLE
YALMIP Time Problem

### DISCUSSION THREAD

**Discussion by phet90** [yalmip/YALMIP#1086] [General discussion] [answered]
Hi everyone,

I have written a function in MATLAB that utilizes YALMIP for structuring the optimization problem, which is a special form of MPC. I use YALMIP so that I can give the objective function and constraints easily. Then I use Gurobi as the solver. This is what I do in steps:
 
1.	I write the MATLAB function that uses YALMIP to construct the optimization problem and choose the latest version of Gurobi as the solver (other solvers like Mosek perform poorly). Let’s call the MATLAB function dummy1.m, 
2.	I use MATLAB library compiler to create C-shared Libraries out of dummy1.m. I get then dummy1.dll, dummy1.lib and dummy1.h from the library compiler,
3.	I use dummy1.h in a different C-code (let’s call it dummy2) and through Visual Studio I create dummy2.dll and dummy2.lib files,
4.	Then I input dummy1.dll, dummy1.lib, dumm2.dll and dummy2.lib into my simulation environment, which has no direct interface with MATLAB. In this way, I am in fact running the “original” MATLAB function (dummy1.m) in my simulation environment.
 
In the original MATLAB function, I save some variables (like sol = optimize(Constraints, Objective, ops)) from the workspace in a specified folder, so that I can track what is happening when the code is executed in the simulation environment (through the DLLs). The problem is that the simulation runs quite slow. I see that the yalmiptime is the one that takes “a lot” of time (around 1-2 seconds) and the solver, which is Gurobi in this case, takes around 0.02 seconds (solvertime out of the sol.struct), picture below.
 
I want to know if there are some tricks, so that I can speed up the executions. I have tried to choose the sdpsettings that save computational effort, but the problem appears again after some minutes of simulations. 
 
Another interesting issue: when the objective function is formulated as norm(something) everything runs smooth. When I change it to norm(something)^2, which in fact is the correct form, this problem happens!  

Would appreciate any inputs/ideas.

![asd](https://user-images.githubusercontent.com/93793919/155884344-9a94e823-f09a-4cc0-9390-841e20dfbdee.jpg)

**Comment by johanlofberg** [comment, marked as answer]
Sounds like you re creating a very bad dense quadratic function, thus getting massive overhead both in symbolic processing and memory, such as minimizing norm(sum(x)) (which is very simple) using a quadratic function (sum(x))^2 which will have O(n^2) monomials and thus have large symbolic overhead and a dense n^2 objective matrix

**Reply by phet90** [reply]
I understand. So it's a matter of formulation.

So there would be no solution except for trying to reformulate the problem?

**Reply by johanlofberg** [reply]
No reason to square it unless it is only part of the objective, and then you could introduce a new variables, i..e instead of norm(stuff)^2 + something, you write it as norm(e)^2 + something and add the constraint e == stuff with a new variables e, thus enforcing a sparse quadratic.

**Reply by phet90** [reply]
Thanks. I understand what you mean. I will try it and let you know how the results are. 

Thanks for responding so good and fast!

**Reply by phet90** [reply]
Thanks Johan, worked nicely.
